pkg/controlplane/controller/leaderelection/leaderelection_controller.go

```diff
@@ -0,0 +1,512 @@
+	if lc.Spec.PingTime != nil {
+		return
+	}
+
+	klog.Infof("reconcile found canLead label namespace=%q, name=%q: %q", lc.Namespace, lc.Name, lc.Spec.LeaseName)
+	leaderLeaseID := leaderLeaseID{lc.Namespace, lc.Spec.LeaseName}
+
+	klog.Infof("Adding %s to electionQueue", leaderLeaseID)
+	c.electionQueue.Add(leaderLeaseID)
+}
+
+func (c *Controller) processLease(obj any) {
+	lease, ok := obj.(*v1.Lease)
+	if !ok {
+		return
+	}
+	if lease != nil {
+		// If the lease is not held by a CLE aware holder and the lease is not expired
+		// then PreferredHolder is not recognized and election is unnecessary since
+		// the holder will not voluntarily give up the lease.
+		// This may occur on version skew cases when CLE is first enabled.
+		if lease.Annotations[ElectedByAnnotationName] != controllerName && !isLeaseExpired(lease) {
+			return
+		}
+	}
+
+	c.electionQueue.Add(leaderLeaseID{namespace: lease.Namespace, name: lease.Name})
+}
+
+func (c *Controller) electionNeeded(candidates []*v1alpha1.LeaseCandidate, leaderLeaseID leaderLeaseID) (bool, error) {
+	lease, err := c.leaseInformer.Lister().Leases(leaderLeaseID.namespace).Get(leaderLeaseID.name)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return false, fmt.Errorf("error reading lease")
+	} else if apierrors.IsNotFound(err) {
+		return true, nil
+	}
+	if isLeaseExpired(lease) || lease.Spec.HolderIdentity == nil || *lease.Spec.HolderIdentity == "" {
+		return true, nil
+	}
+
+	if len(candidates) == 0 {
+		return true, nil
+	}
+
+	prelimStrategy := pickBestStrategy(candidates)
+	if prelimStrategy != v1.OldestEmulationVersion {
+		return false, fmt.Errorf("strategy %s is not recognized by CLE", prelimStrategy)
+	}
+
+	prelimElectee := pickBestLeaderOldestEmulationVersion(candidates)
+	if lease != nil && lease.Spec.HolderIdentity != nil && prelimElectee.Name == *lease.Spec.HolderIdentity {
+		klog.V(2).Infof("Leader %s is already most optimal for lease %s %s", prelimElectee.Name, leaderLeaseID.namespace, leaderLeaseID.name)
+		return false, nil
+	}
+	return true, nil
+}
+
+// reconcileElectionStep steps through a step in an election.
+// A step looks at the current state of Lease and LeaseCandidates and takes one of the following action
+// - do nothing (because leader is already optimal or still waiting for an event)
+// - request ack from candidates (update LeaseCandidate PingTime)
+// - finds the most optimal candidate and elect (update the Lease object)
+// Instead of keeping a map and lock on election, the state is
+// calculated every time by looking at the lease, and set of available candidates.
+// PingTime + electionDuration > time.Now: We just asked all candidates to ack and are still waiting for response
+// PingTime + electionDuration < time.Now: Candidate has not responded within the appropriate PingTime. Continue the election.
+// RenewTime + 5 seconds > time.Now: All candidates acked in the last 5 seconds, continue the election.
```

Comment:
not sure I understand the last line.

Reply:
If all candidates acked within the electionDuration, proceed with picking a leader.